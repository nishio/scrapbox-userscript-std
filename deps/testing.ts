export * from "https://deno.land/std@0.217.0/testing/asserts.ts";
export * from "https://deno.land/std@0.217.0/testing/snapshot.ts";
