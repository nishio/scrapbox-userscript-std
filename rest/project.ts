import {
  isErr,
  mapAsyncForResult,
  mapErrAsyncForResult,
  type Result,
  unwrapOk,
} from "option-t/plain_result";
import type {
  MemberProject,
  NotFoundError,
  NotLoggedInError,
  NotMemberError,
  NotMemberProject,
  ProjectId,
  ProjectResponse,
} from "@cosense/types/rest";
import { cookie } from "./auth.ts";
import { parseHTTPError } from "./parseHTTPError.ts";
import { type HTTPError, responseIntoResult } from "./responseIntoResult.ts";
import type { FetchError } from "./robustFetch.ts";
import { type BaseOptions, setDefaults } from "./options.ts";

export interface GetProject {
  /** Constructs a request for the /api/project/:project endpoint
   *
   * This endpoint retrieves detailed information about a specific project,
   * which can be either a MemberProject or NotMemberProject depending on the user's access level.
   *
   * @param project The project name to retrieve information for
   * @param init Options including connect.sid (session ID) and other configuration
   * @return The constructed request object
   */
  toRequest: (
    project: string,
    options?: BaseOptions,
  ) => Request;

  /** Extracts project JSON data from the API response
   *
   * Processes the API response and extracts the project information.
   * Handles various error cases including NotFoundError, NotMemberError, and NotLoggedInError.
   *
   * @param res The API response object
   * @return A Result containing either project data or an error
   */
  fromResponse: (
    res: Response,
  ) => Promise<Result<MemberProject | NotMemberProject, ProjectError>>;

  (
    project: string,
    options?: BaseOptions,
  ): Promise<
    Result<MemberProject | NotMemberProject, ProjectError | FetchError>
  >;
}

export type ProjectError =
  | NotFoundError
  | NotMemberError
  | NotLoggedInError
  | HTTPError;

const getProject_toRequest: GetProject["toRequest"] = (project, init) => {
  const { sid, hostName } = setDefaults(init ?? {});

  return new Request(
    `https://${hostName}/api/projects/${project}`,
    sid ? { headers: { Cookie: cookie(sid) } } : undefined,
  );
};

const getProject_fromResponse: GetProject["fromResponse"] = async (res) =>
  mapAsyncForResult(
    await mapErrAsyncForResult(
      responseIntoResult(res),
      async (error) =>
        (await parseHTTPError(error, [
          "NotFoundError",
          "NotLoggedInError",
          "NotMemberError",
        ])) ?? error,
    ),
    (res) => res.json() as Promise<MemberProject | NotMemberProject>,
  );

/** get the project information
 *
 * @param project project name to get
 * @param init connect.sid etc.
 */
export const getProject: GetProject = /* @__PURE__ */ (() => {
  const fn: GetProject = async (
    project,
    init,
  ) => {
    const { fetch } = setDefaults(init ?? {});

    const req = getProject_toRequest(project, init);
    const res = await fetch(req);
    if (isErr(res)) return res;

    return getProject_fromResponse(unwrapOk(res));
  };

  fn.toRequest = getProject_toRequest;
  fn.fromResponse = getProject_fromResponse;

  return fn;
})();

export interface ListProjects {
  /** Constructs a request for the /api/projects endpoint
   *
   * This endpoint retrieves information for multiple projects in a single request.
   * The endpoint requires at least one project ID to be provided.
   *
   * @param projectIds Array of project IDs to retrieve information for (must contain at least one ID)
   * @param init Options including connect.sid (session ID) and other configuration
   * @return The constructed request object
   */
  toRequest: (
    projectIds: ProjectId[],
    init?: BaseOptions,
  ) => Request;

  /** Extracts projects JSON data from the API response
   *
   * Processes the API response and extracts information for multiple projects.
   * Handles authentication errors (NotLoggedInError) and other HTTP errors.
   *
   * @param res The API response object
   * @return A Result containing either project data or an error
   */
  fromResponse: (
    res: Response,
  ) => Promise<Result<ProjectResponse, ListProjectsError>>;

  (
    projectIds: ProjectId[],
    init?: BaseOptions,
  ): Promise<Result<ProjectResponse, ListProjectsError | FetchError>>;
}

export type ListProjectsError = NotLoggedInError | HTTPError;

const ListProject_toRequest: ListProjects["toRequest"] = (projectIds, init) => {
  const { sid, hostName } = setDefaults(init ?? {});
  const params = new URLSearchParams(
    projectIds.map((id) => ["ids", id]),
  );

  return new Request(
    `https://${hostName}/api/projects?${params}`,
    sid ? { headers: { Cookie: cookie(sid) } } : undefined,
  );
};

const ListProject_fromResponse: ListProjects["fromResponse"] = async (res) =>
  mapAsyncForResult(
    await mapErrAsyncForResult(
      responseIntoResult(res),
      async (error) =>
        (await parseHTTPError(error, ["NotLoggedInError"])) ?? error,
    ),
    (res) => res.json() as Promise<ProjectResponse>,
  );

/** list the projects' information
 *
 * @param projectIds project ids. This must have more than 1 id
 * @param init connect.sid etc.
 */
export const listProjects: ListProjects = /* @__PURE__ */ (() => {
  const fn: ListProjects = async (
    projectIds,
    init,
  ) => {
    const { fetch } = setDefaults(init ?? {});

    const res = await fetch(ListProject_toRequest(projectIds, init));
    if (isErr(res)) return res;

    return ListProject_fromResponse(unwrapOk(res));
  };

  fn.toRequest = ListProject_toRequest;
  fn.fromResponse = ListProject_fromResponse;

  return fn;
})();
