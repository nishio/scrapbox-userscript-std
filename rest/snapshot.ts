import type {
  ErrorLike,
  NotFoundError,
  NotLoggedInError,
  NotMemberError,
  PageSnapshotList,
  PageSnapshotResult,
} from "@cosense/types/rest";
import { cookie } from "./auth.ts";
import { type BaseOptions, setDefaults } from "./options.ts";
import { parseHTTPError } from "./parseHTTPError.ts";
import {
  isErr,
  mapAsyncForResult,
  mapErrAsyncForResult,
  type Result,
  unwrapOk,
} from "option-t/plain_result";
import { type HTTPError, responseIntoResult } from "./responseIntoResult.ts";
import type { FetchError } from "./mod.ts";

/** Error that occurs when an invalid `timestampId` is provided to {@linkcode getSnapshot} */
export interface InvalidPageSnapshotIdError extends ErrorLike {
  name: "InvalidPageSnapshotIdError";
}

export type SnapshotError =
  | NotFoundError
  | NotLoggedInError
  | NotMemberError
  | InvalidPageSnapshotIdError
  | HTTPError;

/** get a page snapshot
 *
 * @param options `connect.sid` etc.
 */
export const getSnapshot = async (
  project: string,
  pageId: string,
  timestampId: string,
  options?: BaseOptions,
): Promise<Result<PageSnapshotResult, SnapshotError | FetchError>> => {
  const { sid, hostName, fetch } = setDefaults(options ?? {});

  const req = new Request(
    `https://${hostName}/api/page-snapshots/${project}/${pageId}/${timestampId}`,
    sid ? { headers: { Cookie: cookie(sid) } } : undefined,
  );

  const res = await fetch(req);
  if (isErr(res)) return res;

  return mapAsyncForResult(
    await mapErrAsyncForResult(
      responseIntoResult(unwrapOk(res)),
      async (error) =>
        error.response.status === 422
          ? {
            name: "InvalidPageSnapshotIdError",
            message: await error.response.text(),
          }
          : (await parseHTTPError(error, [
            "NotFoundError",
            "NotLoggedInError",
            "NotMemberError",
          ])) ?? error,
    ),
    (res) => res.json() as Promise<PageSnapshotResult>,
  );
};

export type SnapshotTimestampIdsError =
  | NotFoundError
  | NotLoggedInError
  | NotMemberError
  | HTTPError;

/**
 * Retrieves the timestamp IDs for a specific page in a project.
 *
 * @param project - The name of the project.
 * @param pageId - The ID of the page.
 * @param options - Optional configuration options.
 * @returns A promise that resolves to a {@linkcode Result} object containing the page snapshot list if successful,
 * or an error if the request fails.
 */
export const getTimestampIds = async (
  project: string,
  pageId: string,
  options?: BaseOptions,
): Promise<
  Result<PageSnapshotList, SnapshotTimestampIdsError | FetchError>
> => {
  const { sid, hostName, fetch } = setDefaults(options ?? {});

  const req = new Request(
    `https://${hostName}/api/page-snapshots/${project}/${pageId}`,
    sid ? { headers: { Cookie: cookie(sid) } } : undefined,
  );

  const res = await fetch(req);
  if (isErr(res)) return res;

  return mapAsyncForResult(
    await mapErrAsyncForResult(
      responseIntoResult(unwrapOk(res)),
      async (error) =>
        (await parseHTTPError(error, [
          "NotFoundError",
          "NotLoggedInError",
          "NotMemberError",
        ])) ?? error,
    ),
    (res) => res.json() as Promise<PageSnapshotList>,
  );
};
