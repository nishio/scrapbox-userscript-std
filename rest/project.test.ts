import { getProject, listProjects } from "./project.ts";
import { assertSnapshot } from "@std/testing/snapshot";

Deno.test("getProject", async (t) => {
  await assertSnapshot(
    t,
    getProject.toRequest("takker"),
  );
});
Deno.test("listProjects", async (t) => {
  await assertSnapshot(
    t,
    listProjects.toRequest(["dummy-id1", "dummy-id2"]),
  );
});
