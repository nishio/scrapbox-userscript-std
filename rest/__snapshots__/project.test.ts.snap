export const snapshot = {};

snapshot[`getProject 1`] = `
Request {
  bodyUsed: false,
  headers: Headers {},
  method: "GET",
  redirect: "follow",
  url: "https://scrapbox.io/api/projects/takker",
}
`;

snapshot[`listProjects 1`] = `
Request {
  bodyUsed: false,
  headers: Headers {},
  method: "GET",
  redirect: "follow",
  url: "https://scrapbox.io/api/projects?ids=dummy-id1&ids=dummy-id2",
}
`;
